server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -145,69 +145,79 @@ public void start() {
+                                masterOperation(task, request, clusterState, delegate);
+                            }
+                        });
+                    }
                 } else {
-                    DiscoveryNode masterNode = nodes.getMasterNode();
-                    final String actionName = getMasterActionName(masterNode);
-                    transportService.sendRequest(masterNode, actionName, request, new ActionListenerResponseHandler<Response>(listener,
-                        TransportMasterNodeAction.this::newResponse) {
-                        @Override
-                        public void handleException(final TransportException exp) {
-                            Throwable cause = exp.unwrapCause();
-                            if (cause instanceof ConnectTransportException) {
-                                // we want to retry here a bit to see if a new master is elected
-                                logger.debug("connection exception while trying to forward request with action name [{}] to master node [{}], scheduling a retry. Error: [{}]",
+                    if (nodes.getMasterNode() == null) {
+                        logger.debug("no known master node, scheduling a retry");
+                        retry(null, masterChangePredicate);
+                    } else {
+                        DiscoveryNode masterNode = nodes.getMasterNode();
+                        final String actionName = getMasterActionName(masterNode);
+                        transportService.sendRequest(masterNode, actionName, request, new ActionListenerResponseHandler<Response>(listener,
+                            TransportMasterNodeAction.this::newResponse) {
+                            @Override
+                            public void handleException(final TransportException exp) {
+                                Throwable cause = exp.unwrapCause();
+                                if (cause instanceof ConnectTransportException) {
+                                    // we want to retry here a bit to see if a new master is elected
+                                    logger.debug("connection exception while trying to forward request with action name [{}] to master node [{}], scheduling a retry. Error: [{}]",
                                         actionName, nodes.getMasterNode(), exp.getDetailedMessage());
-                                retry(cause, masterChangePredicate);
-                            } else {
-                                listener.onFailure(exp);
+                                    retry(cause, masterChangePredicate);
+                                } else {
+                                    listener.onFailure(exp);
+                                }
                             }
-                        }
-                    });
+                        });
+                    }
                 }
+            } catch (Exception e) {
+                listener.onFailure(e);
```

Comment:
While it might be tricky to see in this diff what's actually changed: The two lines above here have been added.